[RDY]  graph tile splitting

ramirezmike opened the pull request.

ramirezmike commented:
At this point, I think I've reached "feature parity" and am focused on testing it and trying to break it at the moment.

The last couple things I implemented included the margin/padding although, I may have changed the behavior slightly. The existing nog implementation combines margin & padding but I wrote them up to be separate. I'm still testing that piece out a bit;  there are some setups where windows get shifted slightly.

I also added a component for my sanity so I can tell what my next split is or if I'm in fullscreen.

Still planning on adding more unit tests, code documentation, and updating relevant parts of the main documentation and then I guess working on refactoring based on feedback. I'd say it's definitely in a usable state though.

TimUntersberger commented:
> The existing nog implementation combines margin & padding but I wrote them up to be separate.

I had it like this before, but I remember that sometimes you could have rounding issues, which results in uneven padding.

I have very little free time this month, so I might take some time to respond to something. I'll look at the PR this Sunday.

TimUntersberger commented:
I feel like the `cross-platform` branch should be merged before we merge this. The problem is that I refactor the foundation of nog, so now we have to change a fuck ton and this PR will need a lot of changing to merge together with `cross-platform` and I don't even know whether the `cross-platform` branch is working, because of the massive amount of errors I have to fix. Could you help me make the `cross-platform` branch compile so that I can finally try to merge this into `development`? Sorry, I really overdid it with the refactoring. 

On another note I think once this PR and the other one are merged we'll have a lot more possiblities in the future.

ramirezmike commented:
Yeah I took a peak at the cross platform branch and saw the overlap. I agree it makes sense to handle that one first and then I can rebase this on top.

> Could you help me make the cross-platform branch compile so that I can finally try to merge this into development? 

Yeah, I'll take a look. No worries

TimUntersberger commented:
Have you started with the refactor yet?

ramirezmike commented:
> Have you started with the refactor yet?

Yeah, started the rebase. I'll probably have more time this week to get it done. I don't think it'll be too bad so far.

ramirezmike commented:
Small update: still chipping away at this.. I have to just merge some of the rendering differences and then I can get to actually running it regularly to re-test everything. Aiming for an updated pull request by the weekend hopefully of something usable and then going to push more commits on top as I find and fix issues/change to address feedback.

ramirezmike commented:
I'm in the early-testing phase at this point, just ironing out bugs/differences that I need to update.

I have been pulling in the changes you've been pushing to the cross-platform branch, haven't gotten to 01bb20e136e2adc1f74060b2640e5d5174a9c333 yet. 

However, one issue I'm running into that's slowing me down a bit has to do with modes.. I haven't needed to change any of the keybinding stuff really so I'm not clear why it's happening, but when I enter a mode the keybinding to exit the mode isn't registering. Like.. it does get registered but it's like nog doesn't pick up the keybinding at all, it doesn't get to the point where it even logs that a keybinding was entered.


Are you having any problems switching out of modes? @TimUntersberger 

This is what my mode config for resize looks like


    mode "resize" "Alt+Shift+R" {

    bind "H" resize("Left", 2);
    bind "Shift+H" resize("Left", -2);
   
    bind "J" resize("Down", 2);
    bind "Shift+J" resize("Down", -2);

    bind "K" resize("Up", 2);
    bind "Shift+K" resize("Up", -2);

    bind "L" resize("Right", 2);
    bind "Shift+L" resize("Right", -2);

    bind "A" focus("Left");
    bind "W" focus("Up");
    bind "D" focus("Right");
    bind "S" focus("Down");
    }

TimUntersberger commented:
I fixed it in the latest commit.

ramirezmike commented:
Ok, so I updated the pull request. 

There's still going back and adding docs/tests and more comments, but I think it's at a state where it's usable and rebased on latest development branch. I'll continue updating it if I find issues.

If you want to test out the tiling, this is the "mode" file I use

    mode "move" "Alt+M" {
        bind "W" focus("Up");
        bind "A" focus("Left");
        bind "S" focus("Down");
        bind "D" focus("Right");

        bind "H" swap("Left");
        bind "J" swap("Down");
        bind "K" swap("Up");
        bind "L" swap("Right");

        bind "Alt+H" move_in("Left");
        bind "Alt+J" move_in("Down");
        bind "Alt+K" move_in("Up");
        bind "Alt+L" move_in("Right");

        bind "Alt+Shift+H" move_out("Left");
        bind "Alt+Shift+J" move_out("Down");
        bind "Alt+Shift+K" move_out("Up");
        bind "Alt+Shift+L" move_out("Right");
    }

ramirezmike commented:
Ok just a quick recap

Before:

![nog_1](https://user-images.githubusercontent.com/1421719/97113654-31ce4400-16c2-11eb-9a7d-91a46e746d97.JPG)

After:

![nog_2](https://user-images.githubusercontent.com/1421719/97113658-3c88d900-16c2-11eb-8a94-284d6be012fa.JPG)

**Swap** up/down moves a tile up or down within its column. 
**Swap** left/right moves a tile left and right within its row.

(I'm considering adding a _bubbling_ _up_ feature to the swap so that applies the swap to the next parent node that makes sense if the current tile can't complete the operation)

**MoveIn/MoveOut** behaves as follows..

Starting from this layout (notepad | cmd | vim) with the vim tile focused

![nog_3](https://user-images.githubusercontent.com/1421719/97113883-93db7900-16c3-11eb-8053-271347c09a7f.JPG)

**Move_In** **Left** moves the focused vim tile into a row node with the cmd tile

![nog_4](https://user-images.githubusercontent.com/1421719/97113891-9b9b1d80-16c3-11eb-8449-ae1ff6057533.JPG)

**Move_Out** **Up** creates a parent row node with the focused vim tile as the first child and the column node (with the notepad and cmd tiles as children)  as the second child

![nog_5](https://user-images.githubusercontent.com/1421719/97113901-a190fe80-16c3-11eb-9e78-201a8ddaf8ab.JPG)

I did notice I need to update a couple key bindings changes that are causing the grids to lose their focused tile. I would say at this point pulling it down and compiling it is only worth it to check out the tiling on the latest code but otherwise it might be worth waiting until I update with fixes for the focusing issues I noticed.

ramirezmike commented on src/config.rs:
```diff
@@ -113,7 +113,7 @@ impl Config {
             .keybindings
             .iter_mut()
-            .find(|kb| kb.key == keybinding.key && kb.modifier == keybinding.modifier)
+            .find(|kb| kb.key == keybinding.key && kb.modifier == keybinding.modifier && kb.mode == keybinding.mode)
```
this was to fix an issue where keybindings would overwrite keybindings of other modes. Like, there could only exist one binding instance across modes because it wasn't checking if an existing keybinding was for the same mode or a different one

ramirezmike commented on src/display.rs:
```diff
@@ -56,33 +56,11 @@ impl Display {
                 0
             }
     }
-    pub fn get_taskbar_position(&self) -> TaskbarPosition {
```
this was in two places and both of them weren't being used. I removed this one.

ramirezmike commented on src/task_bar.rs:
```diff
@@ -19,10 +18,17 @@ impl Default for TaskbarPosition {
 pub struct Taskbar {
     pub window: NativeWindow,
-    pub position: TaskbarPosition,
+    position: TaskbarPosition,
```
get_poition wasn't getting used and it was causing issues when hiding/showing the taskbar/appbars. I updated the code so that this is private and then all references to the taskbar position would instead use the function get_position

ramirezmike commented on src/hot_reload.rs:
```diff
@@ -83,7 +83,7 @@ pub fn update_config(state_arc: Arc<Mutex<AppState>>, new_config: Config) -> Sys
     for d in state.displays.iter() {
         if let Some(grid) = d.get_focused_grid() {
-            grid.draw_grid(d, &old_config)?;
+            grid.draw_grid(d, &state.config)?;
```
I think this was a bug here... it was reloading the grids with the old configuration instead of the one that just got hot-reloaded. I noticed it because it was causing issues hiding/showing the appbar but once I changed this to use the new config everything started working correctly as far as I can tell.

ramirezmike commented:
Rebased on latest and did a smoke test + running unit tests, seems to all be working well.

Current TODO list

- [ ]  replace Panics with SystemResult usage
- [ ]  add more unit tests and code comments for generated docs
- [ ]  update doc pages with binding information/examples

ramirezmike commented:
I did notice that the current version of nog allows the user to use all keybindings while a tile is fullscreened. I had added checks in several functions that I thought would be confusing if the user accidentally triggered them while fullscreened like swapping tiles or moving in/moving out.

@TimUntersberger Would you prefer if those checks were removed? Another thought I had was to leave the checks in place but allow the user to focus left/right/up/down while fullscreened to switch applications but have the focus behavior ignore the grid layout and instead move forward/backward iterating across the tiles in the grid. That way it ignores the complexity of remembering whether a tile is above or 3 columns over but still allows you to iterate through the whole grid.

TimUntersberger commented:
I quite like the idea of being able to use all of the keybindings when a window is fullscreened. Maybe we could introduce a setting for that?

ramirezmike commented:
> I quite like the idea of being able to use all of the keybindings when a window is fullscreened. Maybe we could introduce a setting for that?

Sure, I can default it to allow keybindings while fullscreened.

ramirezmike commented:
I added a config "ignore_fullscreen_actions" that defaults to false but am still a little unsure on how it should behave when doing different things while the grid has a tile fullscreened. For example, focusing to the right would set focus to a tile that is to the right of the one currently fullscreened... should that change what is currently fullscreened as well?

That aside, have you had a chance to check out some of the other changes?

TimUntersberger commented:
> For example, focusing to the right would set focus to a tile that is to the right of the one currently fullscreened... should that change what is currently fullscreened as well

fullscreen mode only displays the focused tile right now, so that means if you change the focused tile the displayed tile should also change.

> That aside, have you had a chance to check out some of the other changes?

I thought you are still working on this PR. If you feel like the PR is ready please change the title from `[WIP]` to `[RDY]`.

ramirezmike commented:
Cool, yeah. I think it's in a good state to start reviewing and I can make changes as needed. Let me know if you'd prefer it as one squashed commit vs how it is currently.

TimUntersberger commented:
I played with your PR and I don't notice any problems. One thing I am happily surprised by is the speed gain this PR adds. Switching workspaces feels a lot more smooth and fast now.

If you don't have anything else you want to do then please rebase ontop of the latest development commit.

TimUntersberger commented:
I also added you as a collaborator to this repo as you showed that you truly care about this project and your code looks really professional.

ramirezmike commented:
> If you don't have anything else you want to do then please rebase ontop of the latest development commit.

Cool, yeah. I added one small fix for the current_window component just needed to sync it with where the window title comes from now. The pull request is now a set of commits on top of the latest development commit. It should be good to go, but let me know if you're prefer it squashed into one commit instead of separate ones.

>I also added you as a collaborator to this repo as you showed that you truly care about this project and your code looks really professional.

Thank you, I really appreciate that. This project has made my work days a lot more productive and fun and it's cool to be a part of it. I'm excited to see where it goes.
